src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -1593,5 +1598,45 @@ private static void VerifyDerivedTypeConstraints(IEdmModel model, IEdmVocabulary
 
             throw new ODataException(Strings.PathParser_TypeCastOnlyAllowedInDerivedTypeConstraint(fullTypeName, kind, name));
         }
+
+        private static string ResolveEscapeFunction(string identifier, IEdmType bindingType, IEdmModel model, out string parenthesisExpression)
+        {
+            Debug.Assert(identifier != null && identifier.Length >= 1 && identifier[0] == ':');
+            Debug.Assert(bindingType != null);
+
+            IEnumerable<IEdmOperation> operations = model.FindBoundOperations(bindingType);
+            IEdmFunction function = operations.OfType<IEdmFunction>().FirstOrDefault(f => IsUrlEscapeFunction(model, f));
+            if (function == null)
+            {
+                throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_NoBoundEscapeFunctionSupported(bindingType.FullTypeName()));
+            }
+
+            if (function.Parameters == null || function.Parameters.Count() != 2 || !function.Parameters.ElementAt(1).Type.IsString())
+            {
+                throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_EscapeFunctionMustHaveOneStringParameter(function.FullName()));
+            }

```

Comment:
if we can't resolve the function as an escape function, would it be more backward compatible just to return identifier? What would have happened previously if a segment started with a colon? Would this ever have been legal previously?

Reply:
i.e., should we make this TryResolveEscapeFunction and, if it fails, continue existing processing? Or are we sure existing request processing would fail, so we might as well fail here.

---
In reply to: [255710193](https://github.com/OData/odata.net/pull/1400#discussion_r255710193) [](ancestors = 255710193)

Reply:
Note: i suspect that the parse would fail if it contained this syntax, so I'm okay not making this TryResolveEscapeFunction and just failing, I just wanted to be sure we weren't making something that was previously valid now an error.

---
In reply to: [267887753](https://github.com/OData/odata.net/pull/1400#discussion_r267887753) [](ancestors = 267887753,255710193)

Reply:
I just tested:

a: "/Customers(2):/abc:xyz:" and "Customers(2)/MyOrders:/xyz/abc:", it returns "Bad Request - Error in query syntax.".

b: "/root:/photos%2F2018%2F/:February:/Name", it returns "Resource not found for the segment ':photos%2F2018%2F/:February:'."

So, I think to throw the better exception message in "TryResolveEscapeFunction" is good. 




---
In reply to: [268329867](https://github.com/OData/odata.net/pull/1400#discussion_r268329867) [](ancestors = 268329867,267887753,255710193)